### PR TITLE
More Main Menu Content

### PR DESCRIPTION
**Pull request by Jade-Harleyy**
This PR adds:
- Buttons for the remaining content editors to the main menu.
- Singleplayer mission/sandbox modes (latter accessible by selecting no available missions for the former).

image

### **New Localization Tags:**
- `leveleditorbutton` - Level Editor
- `particleeditorbutton` - Particle Editor
- `eventeditorbutton` - Event Editor
- `spriteeditorbutton` - Sprite Editor
- `newmissionbutton` - Start Mission

### **Modified Localization Tags:**
- `campaignlabel` - Campaign => Singleplayer
- `newgamebutton` - New Game => New Campaign
- `loadgamebutton` - Load Game => Load Campaign

### CODE REVIEW

**Comment by EebleSheembl** [comment]
Need

**Comment by PanmanS** [comment]
This would significantly simplify testing gor specific missions when `quickstart` doesnt suffice and serve lobby is too long to load

**Comment by Regalis11** [comment]
> Buttons for the remaining content editors to the main menu.

I don't think all of these should be included in the main menu. Many of them are not intended to be used the average player, but rather advanced modders who are expected to be able to access them using console commands. If we were to make them visible in the main menu, we would need to localize all the texts in them to all the languages the game supports, and preferably also make them a lot more user-friendly.

I also feel a little iffy about including the mission mode in the main menu like this. The mission mode is so underdeveloped that in my opinion it needs to be reworked before it's put into the spotlight like this.

**Comment by NotWendy** [comment]
When I was getting into the game around 1.0, I used the mission mode as a training tool, to understand the very basics of the game, in addition to doing some of the tutorials. It may be underdeveloped right now, but it still works well enough for that purpose, allowing a new player to learn the ropes and fail in a low stakes environment.
I understand you don't want players to get the wrong impression of what the mission mode is supposed to be by trying it out before its big overhaul happens. But with the value it already brings to the game, I think you can consider adding it to the main menu early.

**Comment by WrillWasTaken** [comment]
> This would significantly simplify testing for specific missions when `quickstart` doesnt suffice and serve lobby is too long to load

wouldn't the better solution be to just add difficulty, bot count and biome to the quickstart command?

**Review by Regalis11 — CHANGES_REQUESTED** [review]
While the rest of the changes look good to me, I don't think we can merge this in it's current form

The "hidden editors" should not be included in the main menu. Many of them are not intended to be used the average player, but rather advanced modders who are expected to be able to access them using console commands. If we were to make them visible in the main menu, we would need to localize all the texts in them to all the languages the game supports, and preferably also make them a lot more user-friendly.

File: Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs [hunk 1]
```diff
@@ -35,14 +35,20 @@ private enum Tab
             SubmarineEditor = 7,
             Mods = 8,
             Credits = 9,
-            Empty = 10
+            Empty = 10,
+            LevelEditor = 11,
+            ParticleEditor = 12,
+            EventEditor = 13,
+            SpriteEditor = 14,
```
Comment (Regalis11) [inline Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs:42]: The above enums are unused

File: Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs [hunk 2]
```diff
@@ -391,27 +408,49 @@ GUITextBlock createRightText(LocalizedString str)
...
+                OnClicked = (tb, userdata) => SelectTab(tb, userdata)
+            };
+
+            new GUIButton(new RectTransform(Vector2.One, customizeList.RectTransform), TextManager.Get("SpriteEditorButton"), textAlignment: Alignment.Left, style: "MainMenuGUIButton")
+            {
+                ForceUpperCase = ForceUpperCase.Yes,
+                UserData = Tab.SpriteEditor,
+                OnClicked = (tb, userdata) => SelectTab(tb, userdata)
             };
+            */
```
Comment (Regalis11) [inline Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs:453]: Commented out code should be removed

File: Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs [hunk 3]
```diff
@@ -817,6 +858,29 @@ var newServerExes
...
+                    break;
+                case Tab.ParticleEditor:
+                    CoroutineManager.StartCoroutine(SelectScreenWithWaitCursor(GameMain.ParticleEditorScreen));
+                    break;
+                case Tab.EventEditor:
+                    CoroutineManager.StartCoroutine(SelectScreenWithWaitCursor(GameMain.EventEditorScreen));
+                    break;
+                case Tab.SpriteEditor:
+                    CoroutineManager.StartCoroutine(SelectScreenWithWaitCursor(GameMain.SpriteEditorScreen));
+                    break;
```
Comment (Regalis11) [inline Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs:872]: These are not (and should not be) used

File: Barotrauma/BarotraumaClient/ClientSource/Screens/SinglePlayerMissionSetupUI.cs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+#endif
+            }
+
+            if (subsToShow.Where(s => s.IsCampaignCompatible).ToList() is { Count: > 0 } validSubs)
+            {
+                subList.Select(validSubs.GetRandomUnsynced());
+            }
+        }
+    }
+}
```
Comment (Regalis11) [inline Barotrauma/BarotraumaClient/ClientSource/Screens/SinglePlayerMissionSetupUI.cs:319]: There is a huge amount of code duplication between this class and CampaignSetupUI. The shared logic should most likely be moved to a shared base class.

File: Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs [hunk 5]
```diff
@@ -315,12 +315,17 @@ protected override void UpdateMissionSpecific(float deltaTime)
...
-#if SERVER
-                    if (!(GameMain.GameSession.GameMode is CampaignMode) && GameMain.Server != null)
+                    if (GameMain.GameSession.GameMode is not CampaignMode)
                     {
-                        GameMain.Server.EndGame();                        
-                    }
+#if SERVER
+                        GameMain.Server?.EndGame();
+#else
+                        if (GameMain.GameSession.GameMode is { IsSinglePlayer: true } && GameMain.GameSession.CrewManager is { IsSinglePlayer: true })
```
Comment (Regalis11) [inline Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs:323]: I'm not sure I understand why `IsSinglePlayer: true` is checked on both the GameMode and the CrewManager. Is there a situation in which one is singleplayer, and the other isn't?

File: Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs [hunk 6]
```diff
@@ -335,18 +340,23 @@ protected override void UpdateMissionSpecific(float deltaTime)
...
-                    if (!(GameMain.GameSession.GameMode is CampaignMode) && GameMain.Server != null)
+                    if (GameMain.GameSession.GameMode is not CampaignMode)
                     {
                         if (!Submarine.MainSub.AtStartExit || (wasDocked && !Submarine.MainSub.DockedTo.Contains(Level.Loaded.StartOutpost)))
                         {
-                            GameMain.Server.EndGame();
+#if SERVER
+                            GameMain.Server?.EndGame();
+#else
+                            if (GameMain.GameSession.GameMode is { IsSinglePlayer: true } && GameMain.GameSession.CrewManager is { IsSinglePlayer: true })
```
Comment (Regalis11) [inline Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs:351]: Same here, why does `IsSinglePlayer: true` need to be checked on both of these?

File: Barotrauma/BarotraumaShared/SharedSource/GameSession/GameModes/CoOpMode.cs [hunk 7]
```diff
@@ -1,12 +1,91 @@
...
+            else
+            {
+                endRoundTimer = 0;
+            }
+
+            if (endRoundTimer >= endRoundDelay)
+            {
+                End();
+            }
+        }
```
Comment (Regalis11) [inline Barotrauma/BarotraumaShared/SharedSource/GameSession/GameModes/CoOpMode.cs:63]: There is a ton of code duplication between this class and `GameServer.cs` which handles ending the round in multiplayer.

File: Barotrauma/BarotraumaShared/SharedSource/GameSession/GameModes/CoOpMode.cs [hunk 8]
```diff
@@ -1,12 +1,91 @@
...
-    class CoOpMode : MissionMode
+    internal class CoOpMode : MissionMode
     {
+        private float endRoundTimer;
+
         public CoOpMode(GameModePreset preset, IEnumerable<MissionPrefab> missionPrefabs) : base(preset, ValidateMissionPrefabs(missionPrefabs, MissionPrefab.CoOpMissionClasses)) { }
 
         public CoOpMode(GameModePreset preset, MissionType missionType, string seed) : base(preset, ValidateMissionType(missionType, MissionPrefab.CoOpMissionClasses), seed) { }
+
+#if CLIENT
```
Comment (Regalis11) [inline Barotrauma/BarotraumaShared/SharedSource/GameSession/GameModes/CoOpMode.cs:15]: It'd be better to move the client-specific code to a partial class.

**Review by Regalis11 — CHANGES_REQUESTED** [review]
I'm still of the opinion that the mission mode is so underdeveloped it should not be put into spotlight like this (first option in the main menu after the tutorials - I think a new player who assumes such a prominently displayed game mode is something they should start with would get a bad first impression of the game).

We have some plans for overhauling it in the future (some of which we're planning to do alongside the PvP overhaul), with things like options to select upgrades and extra supplies, and to select more than one mission per round. Those will most likely require quite a lot of changes to the "setup ui" and the code, so I feel it's not a good idea to include this sort of a "intermediate" version at this stage.

I feel that there's also quite a lot of issues in the code currently: a lot of copypasted, duplicate code, and dead code that does nothing.

File: Barotrauma/BarotraumaShared/SharedSource/GameSession/GameModes/CoOpMode.cs (hunk 8 above)
Comment (Jade-Harleyy) [reply Barotrauma/BarotraumaShared/SharedSource/GameSession/GameModes/CoOpMode.cs]: Fixed.

File: Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs (hunk 5 above)
Comment (Jade-Harleyy) [reply Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs]: Fixed, it now only checks if the GameMode has IsSinglePlayer set. Also fixed in CoOpMode.cs.

File: Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs (hunk 6 above)
Comment (Jade-Harleyy) [reply Barotrauma/BarotraumaShared/SharedSource/Events/Missions/AbandonedOutpostMission.cs]: Fixed; see previous comment.

File: Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs (hunk 2 above)
Comment (Jade-Harleyy) [reply Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs]: Done.

File: Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs (hunk 1 above)
Comment (Jade-Harleyy) [reply Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs]: Fixed.

File: Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs (hunk 3 above)
Comment (Jade-Harleyy) [reply Barotrauma/BarotraumaClient/ClientSource/Screens/MainMenuScreen/MainMenuScreen.cs]: Fixed.

**Comment by Regalis11** [comment]
Closing for the reasons mentioned above. I still think this could come in useful in the future, but for the time being we have no plans of highlighting the mission mode this prominently in the main menu. 

The code also still requires some more work: there's a lot of conflicts, the code quality still needs some improvements, and this would also require a significant amount of testing because it does such extensive changes to pretty fundamental parts of the game (the game mode classes and how they handle things such as ending rounds). Thank you for the contribution regardless!

**Pull request closed without merging**